tests/scenario/test_middlewares.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import unittest
+from unittest.mock import MagicMock, PropertyMock, patch
+
+from charm import TraefikIngressCharm
+from scenario import Container, Context, State, Relation
+
+
+@patch("charm.KubernetesServicePatch")
+@patch("lightkube.core.client.GenericSyncClient")
+class TestMiddlewares(unittest.TestCase):
+    def setUp(self) -> None:
+        self.containers = [Container(name="traefik", can_connect=True)]
+
+        version_patcher = patch.object(TraefikIngressCharm, "version", property(lambda *_: "0.0.0"))
+        self.version_patch = version_patcher.start()
+        self.addCleanup(version_patcher.stop)
+
+    @patch("charm.TraefikIngressCharm.external_host", PropertyMock(return_value="testhostname"))
+    @patch("charm.TraefikIngressCharm._traefik_service_running", PropertyMock(return_value=True))
```

Comment:
you might get rid of this by declaring the service in the Container object, but this is quicker.

Reply:
Could you please elaborate/example link?

Reply:
for example here: https://github.com/canonical/traefik-k8s-operator/pull/180/files#diff-965c0ce47f14d31c34286d6286f9c7a08375cce8ad74ade0a2c274fb6a9affd9R55 I set up the traefik container so that _traefik_service_running returns True. (without needing to mock it).

See this test https://github.com/canonical/traefik-k8s-operator/pull/180/files#diff-965c0ce47f14d31c34286d6286f9c7a08375cce8ad74ade0a2c274fb6a9affd9R62 for usage